### ISSUE TITLE
Segfault in uv__async_io

### ISSUE THREAD

**Issue by saghul** [joyent/libuv#603]
I'm getting some weird crash with latests master. Seems race-y because I can't reproduce all the time :-S

Here is the gdb backtrace:

```
#0  0x0000000000000031 in ?? ()
#1  0x00007fc084683aec in uv__async_io (loop=0x7fc0848e0b40, handle=0x7fc0848e0c78, events=1) at src/unix/async.c:175
#2  0x00007fc084684c28 in uv__io_rw (ev=0x7fc0848e2000, w=0x7fc0848e0c78, events=1) at src/unix/core.c:549
#3  0x00007fc0846a137e in ev_invoke_pending (loop=0x7fc0848e2000) at src/unix/ev/ev.c:2145
#4  0x00007fc0846846b7 in uv__poll (loop=0x7fc0848e0b40) at src/unix/core.c:270
#5  0x00007fc084684701 in uv__run (loop=0x7fc0848e0b40) at src/unix/core.c:279
#6  0x00007fc084684760 in uv_run (loop=0x7fc0848e0b40) at src/unix/core.c:287
#7  0x00007fc0846792ba in Loop_func_run (self=0x186b788) at src/loop.c:54
```

If I inspect the actual handle in the 1st frame:

```
(gdb) f 1
#1  0x00007fc084683aec in uv__async_io (loop=0x7fc0848e0b40, handle=0x7fc0848e0c78, events=1) at src/unix/async.c:175
175             ASYNC_CB(h)
(gdb) list
170
171         /* If we need to sweep all handles anyway - skip this loop */
172         if (!loop->async_sweep_needed) {
173           for (i = 0; i < end; i += sizeof(h)) {
174             h = *((uv_async_t**) (buf + i));
175             ASYNC_CB(h)
176           }
177         }
178
179         bytes -= end;
(gdb) print *h
$1 = {close_cb = 0x184e1b0, data = 0x18d9520, loop = 0x7fc0848e0b40, type = 49, handle_queue = {prev = 0x18dae10, next = 0x7860c0}, flags = 32, next_closing = 0x1863b40, pending = 0, async_cb = 0x31, queue = {prev = 0x18dae50, 
    next = 0x7860c0}}
(gdb) 
```

By looking at this, I can't figure out what is going on. However, the fact that close_cb is set sugests me that uv_close was called, but somehow the flag test in ASYC_CB passed?!

Any idea?

**Comment by indutny** [comment]
Idea - yes, solution - no so far

**Comment by bnoordhuis** [comment]
"Fixed" in b7f38b1 by reverting 209abba.
